test/gurka/test_route.cc

```diff
@@ -416,3 +416,43 @@ TEST(Standalone, LegWeightRegression) {
   // "weight" was negative due to failing to properly update elapsed_cost.cost
   EXPECT_GT(doc["routes"][0]["legs"][0]["steps"][2]["weight"].GetDouble(), 0);
 }
+
+TEST(Standalone, DontIgnoreRestriction) {
+  const std::string ascii_map = R"(
+                D
+               /
+      A----B--C
+           |
+           E
+)";
+
+  const gurka::ways ways = {
+      {"DC", {{"highway", "secondary"}, {"oneway", "yes"}}},
+      {"CB", {{"highway", "secondary"}, {"oneway", "yes"}}},
+      {"BE", {{"highway", "primary"}, {"oneway", "yes"}}},
+      {"BA", {{"highway", "primary"}, {"oneway", "yes"}}},
+  };
+
+  const gurka::relations relations = {
+      {{
+           {gurka::way_member, "DC", "from"},
+           {gurka::way_member, "CB", "via"},
+           {gurka::way_member, "BE", "to"},
+       },
+       {
+           {"type", "restriction"},
+           {"restriction", "no_entry"},
+       }},
+  };
+
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 100);
+  auto map = gurka::buildtiles(layout, ways, {}, relations, "test/data/dont_ignore_restriction",
+                               {{"mjolnir.concurrency", "1"}});
+
+  try {
+    auto result = gurka::route(map, "D", "E", "auto");
+    gurka::assert::raw::expect_path(result, {"Unexpected path found"});
```

Comment:
It returns `DC-CB-BE`

Reply:
From my view the bug is around this line https://github.com/valhalla/valhalla/blob/ff25b1116659d824ce5e1ecb3784b4ffe0a1e43e/src/thor/bidirectional_astar.cc#L764
`rev_pred` == `DC` and `rev_pred.on_complex_restr()` returns false

Reply:
I guess the problem is in [restrictionbuilder.cc](https://github.com/valhalla/valhalla/blob/93c955f78e1a0c61ca7898ef570d771aee9f03ce/src/mjolnir/restrictionbuilder.cc#L319). 

Let's say we have a restriction `A -> B -> C`. So we take the edge `AB` and it's marked as `start_restriction`. We go forward and reach the node `C`. We go backwards and finally have the edges `{CB, BA}`. After that we add reverse restriction using the `BA` edge as from. **BUT** we add it to the tile of the `AB` edge.
tldr: if nodes `A` and `B` belong to different tiles (Houston) we got a problem

Reply:
the good news is this should be incredibly rare! but also wow, good find, what a needle in a haystack. i suppose we have to detect when restrictions cross tile boundaries in the restrictions builder and patch it to handle that case.

Reply:
> the good news is this should be incredibly rare
agreed!